### PR TITLE
Support a package SILLinkage

### PR DESCRIPTION
**Pull request by elsh**
Decls with a package access level are currently set to public SIL linkages. This limits the ability to have more fine-grained control and optimize around resilience and serialization.
This PR introduces a separate SIL linkage and FormalLinkage for package decls, pipes them down to IRGen, and updates linkage checks to include package linkage.

Resolves rdar://121409846

### CODE REVIEW

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

File: lib/SIL/IR/SILModule.cpp [hunk 1]
```diff
@@ -75,12 +75,19 @@ class SILModule::SerializationCallback final
       // translation units in the same Swift module.
       decl->setLinkage(SILLinkage::Shared);
       return;
+    case SILLinkage::Package:
+      decl->setLinkage(SILLinkage::PackageExternal);
+      return;
+    case SILLinkage::PackageNonABI: // Q: Same as PublicNonABI?
+      decl->setLinkage(SILLinkage::Shared);
```
Comment (elsh) [inline lib/SIL/IR/SILModule.cpp:82]: Is it safe to return `shared` here?

File: lib/SIL/IR/SILPrinter.cpp [hunk 2]
```diff
@@ -3215,10 +3215,13 @@ static StringRef getLinkageString(SILLinkage linkage) {
   switch (linkage) {
   case SILLinkage::Public: return "public ";
   case SILLinkage::PublicNonABI: return "non_abi ";
+  case SILLinkage::Package: return "package ";
+  case SILLinkage::PackageNonABI: return "non_abi "; // Q: return non_abi for package nonabi?
```
Comment (elsh) [inline lib/SIL/IR/SILPrinter.cpp:3219]: is it safe to return the same value as L3217 here? @eeckstein @slavapestov

File: lib/SIL/IR/SILModule.cpp (hunk 1 above)
Comment (eeckstein) [reply lib/SIL/IR/SILModule.cpp:82]: I think so

File: lib/SIL/IR/SILPrinter.cpp (hunk 2 above)
Comment (eeckstein) [reply lib/SIL/IR/SILPrinter.cpp:3219]: The textual SIL should represent the internal SIL data structure. Printing `non_abi` for `PackageNonABI` would hide this information in textual SIL. So you should return `package_non_abi`. And also handle this in the SILParser.

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

File: test/Sema/package_resilience_access.swift [hunk 3]
```diff
@@ -258,24 +345,63 @@ package struct PkgStructWithPublicMember {
...
-    var data: Int
-}
 
-// UTILS-RES-LABEL: sil_vtable PublicKlass {
-// UTILS-NONRES-LABEL: sil_vtable [serialized] PublicKlass {
-// UTILS-COMMON-LABEL: sil_vtable PkgKlass {
+// UTILS-RES-DAG: sil_vtable PublicKlass {
+// UTILS-NONRES-DAG: sil_vtable [serialized] PublicKlass {
+// FIXME: need to serialize package if non-resiliently built
+// UTILS-COMMON-DAG: sil_vtable PkgKlass {
```
Comment (elsh) [inline test/Sema/package_resilience_access.swift:359]: RES: sil_vtable UfiPkgKlass
NONRES: sil_vtable [serialized] UfiPkgKlass

File: test/SILGen/accessibility_warnings.swift [hunk 4]
```diff
@@ -141,19 +141,19 @@ internal class InternalClass {
 }
 
 package class PackageClass {
-  // CHECK-DAG: sil{{( \[.+\])*}} [ossa] @$s22accessibility_warnings12PackageClassC9publicVarSivg
+  // CHECK-DAG: sil package [transparent] [ossa] @$s22accessibility_warnings12PackageClassC9publicVarSivg
   public var publicVar = 0
```
Comment (elsh) [inline test/SILGen/accessibility_warnings.swift:145]: 1. add library evolution - getter shouldn't be transparent 
2. compare with public

File: test/Sema/package_resilience_access.swift [hunk 5]
```diff
@@ -45,211 +45,298 @@ public protocol PublicProto {
...
+  // UTILS-COMMON-DAG: sil shared [thunk] [ossa] @$s5Utils8PkgKlassC4dataSivpACTK : $@convention(keypath_accessor_getter) (@in_guaranteed PkgKlass) -> @out Int {
 
+  // UTILS-COMMON-DAG: // key path setter for PkgKlass.data : PkgKlass
+  // UTILS-COMMON-DAG: sil shared [thunk] [ossa] @$s5Utils8PkgKlassC4dataSivpACTk : $@convention(keypath_accessor_setter) (@in_guaranteed Int, @in_guaranteed PkgKlass) -> () {
 
-// UTILS-RES-LABEL: // PkgKlass.data.modify
-// UTILS-RES-NEXT: sil [ossa] @$s5Utils8PkgKlassC4dataSivM : $@yield_once @convention(method) (@guaranteed PkgKlass) -> @yields @inout Int {
-// UTILS-NONRES-LABEL: // PkgKlass.data.modify
-// UTILS-NONRES-NEXT: sil [transparent] [ossa] @$s5Utils8PkgKlassC4dataSivM : $@yield_once @convention(method) (@guaranteed PkgKlass) -> @yields @inout Int {
+  // UTILS-COMMON-DAG: // protocol witness for PkgProto.data.getter in conformance PkgKlass
```
Comment (elsh) [inline test/Sema/package_resilience_access.swift:121]: remove comment

**Comment by elsh** [comment]
@swift-ci test source compatibility

File: include/swift/SIL/SILLinkage.h [hunk 6]
```diff
@@ -198,17 +223,59 @@ inline SILLinkage addExternalToLinkage(SILLinkage linkage) {
...
+  llvm_unreachable("Unhandled SILLinkage in switch.");
 }
 
 /// Return whether the given linkage indicates that an object's
 /// definition might be required outside the current SILModule.
 /// If \p is true then we are in whole-module compilation.
 inline bool isPossiblyUsedExternally(SILLinkage linkage, bool wholeModule) {
   if (wholeModule) {
-    return linkage <= SILLinkage::PublicNonABI;
+    switch (linkage) {
```
Comment (eeckstein) [inline include/swift/SIL/SILLinkage.h:248]: IMO, it would be clearer if you do a single switch and check `wholeModule` only in the `Hidden` case.

**Review by eeckstein — COMMENTED** [review]
The SIL part LGTM!

**Comment by elsh** [comment]
@swift-ci test source compatibility

**Comment by elsh** [comment]
@swift-ci  test source compatibility

**Comment by elsh** [comment]
@swift-ci test source compatibility

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
CI source compat suite failure below is a known issue on main branch (https://github.com/apple/swift/issues/71423) / rdar://122427719 and is unrelated to this PR. 

```
========================================
XFailures:
  XFAIL: https://github.com/apple/swift/issues/70742, CoreStore, 4.2, 286360, CoreStore watchOS, generic/platform=watchOS
  XFAIL: https://github.com/apple/swift/issues/70742, CoreStore, 4.2, 286360, CoreStore tvOS, generic/platform=tvOS
  XFAIL: https://github.com/apple/swift/issues/70742, CoreStore, 4.2, 286360, CoreStore OSX, generic/platform=macOS
  XFAIL: https://github.com/apple/swift/issues/70742, CoreStore, 4.2, 286360, CoreStore iOS, generic/platform=iOS
  XFAIL: https://github.com/apple/swift/issues/70742, CoreStore, 4.0, 83e608, CoreStore watchOS, generic/platform=watchOS
  XFAIL: https://github.com/apple/swift/issues/70742, CoreStore, 4.0, 83e608, CoreStore tvOS, generic/platform=tvOS
  XFAIL: https://github.com/apple/swift/issues/70742, CoreStore, 4.0, 83e608, CoreStore OSX, generic/platform=macOS
  XFAIL: https://github.com/apple/swift/issues/70742, CoreStore, 4.0, 83e608, CoreStore iOS, generic/platform=iOS
  XFAIL: https://github.com/apple/swift/issues/69410, Kingfisher, 5.0, 44450a, Kingfisher, generic/platform=watchOS
  XFAIL: https://github.com/apple/swift/issues/65197, ProcedureKit, 5.0, a7fa56, ProcedureKit, generic/platform=watchOS
  XFAIL: https://github.com/apple/swift/issues/65197, ProcedureKit, 4.2, 94193d, ProcedureKit, generic/platform=watchOS
  XFAIL: https://github.com/apple/swift/issues/65197, PromiseKit, 5.0, 8e5f5d, PromiseKit, generic/platform=watchOS
  XFAIL: https://github.com/apple/swift/issues/65197, PromiseKit, 4.0, c538a2, PromiseKit, generic/platform=watchOS
  XFAIL: https://github.com/apple/swift/issues/65204, R.swift, 5.0, f96758, Swift Package
  XFAIL: https://github.com/apple/swift/issues/65204, R.swift, 4.2, f96758, Swift Package
  XFAIL: https://github.com/apple/swift/issues/70744, ReactiveSwift, 5.0, e60a8b, ReactiveSwift-watchOS, generic/platform=watchOS
  XFAIL: https://github.com/apple/swift/issues/70744, ReactiveSwift, 4.0, 46fb4d, ReactiveSwift-watchOS, generic/platform=watchOS
  XFAIL: https://github.com/apple/swift/issues/70970, Result, 5.0, c30700, Swift Package
  XFAIL: https://github.com/apple/swift/issues/65202, swift-protobuf-plugin-example, 5.0, 85bec0, Swift Package
  XFAIL: https://github.com/apple/swift/issues/65197, mapper, 5.0, 877dc5, Mapper, generic/platform=watchOS
  XFAIL: https://github.com/apple/swift/issues/65197, mapper, 4.2, d501cb, Mapper, generic/platform=watchOS
  XFAIL: https://github.com/apple/swift/issues/65197, mapper, 4.0, d501cb, Mapper, generic/platform=watchOS
  XFAIL: https://github.com/apple/swift/issues/57431, siesta, 5.0, cb9c1b, Swift Package
========================================
Failures:
  FAIL: penny-bot, 5.9, 62006b, Swift Package
========================================
Action Summary:
     Passed: 346
     Failed: 1
    XFailed: 23
    UPassed: 0
      Total: 370
========================================
Repository Summary:
      Total: 123
========================================
Result: FAIL
========================================
```  
https://ci.swift.org/job/swift-PR-source-compat-suite-debug-macos/1378/console
https://ci.swift.org/job/swift-PR-source-compat-suite-macos/1383/console

**Pull request merged**